zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/HdfsNotebookRepo.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.hadoop.io.IOUtils;
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.notebook.Note;
+import org.apache.zeppelin.notebook.NoteInfo;
+import org.apache.zeppelin.user.AuthenticationInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * NotebookRepos for hdfs.
+ *
+ * Assume the notebook directory structure is as following
+ * - notebookdir
+ *              - noteId/note.json
+ *              - noteId/note.json
+ *              - noteId/note.json
+ */
+public class HdfsNotebookRepo implements NotebookRepo {
+  private static final Logger LOGGER = LoggerFactory.getLogger(HdfsNotebookRepo.class);
+
+  private Configuration hadoopConf;
+  private ZeppelinConfiguration zConf;
+  private FileSystem fs;
+  private Path notebookDir;
+
+  public HdfsNotebookRepo(ZeppelinConfiguration zConf) throws IOException {
+    this.zConf = zConf;
+    this.hadoopConf = new Configuration();
+    this.fs = FileSystem.get(new Configuration());
+    this.notebookDir = new Path(zConf.getNotebookDir());
+    if (!fs.exists(notebookDir)) {
+      fs.mkdirs(notebookDir);
+      LOGGER.info("Create notebook dir {} in hdfs", notebookDir.toString());
+    }
+    if (fs.isFile(notebookDir)) {
+      throw new IOException("notebookDir {} is file instead of directory, please remove it or " +
+          "specify another directory");
+    }
+    LOGGER.info("Use hdfs directory {} to store notebook", notebookDir);
+  }
+
+  @Override
+  public List<NoteInfo> list(AuthenticationInfo subject) throws IOException {
+    List<NoteInfo> noteInfos = new ArrayList<>();
+    for (FileStatus status : fs.globStatus(new Path(notebookDir, "*/note.json"))) {
+      NoteInfo noteInfo = new NoteInfo(status.getPath().getParent().getName(), "", null);
+      noteInfos.add(noteInfo);
+    }
+    return noteInfos;
+  }
+
+  @Override
+  public Note get(String noteId, AuthenticationInfo subject) throws IOException {
+    Path notePath = new Path(notebookDir.toString() + "/" + noteId + "/note.json");
+    ByteArrayOutputStream noteBytes = new ByteArrayOutputStream();
+    IOUtils.copyBytes(fs.open(notePath), noteBytes, hadoopConf);
+    return Note.fromJson(new String(noteBytes.toString("UTF-8")));
+  }
+
+  @Override
+  public void save(Note note, AuthenticationInfo subject) throws IOException {
+    Path tmpNotePath = new Path(notebookDir.toString() + "/" + note.getId() + "/.note.json");
+    if (fs.exists(tmpNotePath)) {
+      fs.delete(tmpNotePath, true);
+    }
+    InputStream in = new ByteArrayInputStream(note.toJson().getBytes("UTF-8"));
+    IOUtils.copyBytes(in, fs.create(tmpNotePath), hadoopConf);
+    Path notePath = new Path(notebookDir.toString() + "/" + note.getId() + "/note.json");
+    fs.delete(notePath);
```

Comment:
I am afraid this is what we only can do for now. Each time when we save a note, it would overwrite the previous one. And actually you don't know the file is the previous note or other file.